### PR TITLE
Convert `ag` to rzshell

### PR DESCRIPTION
**Pull request by b1llow**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Before many ag commands had a "format" as a part of the command itself, e.g. agfk - here command was agf while k is the format. For the time being it will become agf k, unless you have better ideas.

`ag<graphtype><format>` -> `ag<graphtype> <format>`

but `ag<graphtype>w` -> `agw <graphtype>`

```
Usage: ag<?>   # Analysis graph commands
| aga <format>=           # Data reference graph
| agA <format>=           # Global data references graph
| agc <format>=           # Function callgraph
| agC <format>=           # Global callgraph
| agd [<format> [<addr>]] # Diff graph
| agf <format>=           # Basic blocks function graph
| agi <format>=           # Imports graph
| agr <format>=           # References graph
| agR <format>=           # Global references graph
| ags                     # Unknown graph
| agl                     # Line graph ?
| agx <format>=           # Cross-references graph
| agg <format>=           # Custom graph
| ag-                     # Clear the custom graph
| agn[-]                  # Managing custom graph nodes
| age[-]                  # Managing custom graph edges
| agw <graphtype> <path>  # Write to path or display graph image (see graph.gv.format)
```

Maybe open another PR to do more refactoring and export some APIs
But this PR mainly focuses on convert to rzshell

Test plan

CI is green

Closing issues

Partially addresses https://github.com/rizinorg/rizin/issues/1442

previous PR https://github.com/rizinorg/rizin/pull/2757

### CODE REVIEW

File: librz/core/cmd_descs/cmd_analysis.yaml
Comment (XVilka) [inline librz/core/cmd_descs/cmd_analysis.yaml]: Maybe make it `"normal"`? And have the default format called "normal". @ret2libc what do you think?

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cmd_descs/cmd_analysis.yaml
Comment (XVilka) [reply librz/core/cmd_descs/cmd_analysis.yaml]: @ret2libc is it possible to make a set of strings as the only possible values and autocomplete them?

File: librz/core/cmd/cmd_analysis.c
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c]: I think these comments are obsolete and should be removed. Even case name provides better meaning for the purpose of this code.

**Review by XVilka — CHANGES_REQUESTED** [review]
Assuming it's only the first step and there will be some more additional cleanup and refactoring, this PR is OK except for one thing, in my opinion, - the `default_value: ' '` of the format. I hope there could be a better solution. cc @ret2libc @wargio

File: librz/core/cmd/cmd_analysis.c
Comment (XVilka) [inline librz/core/cmd/cmd_analysis.c]: Shouldn't we remove this completely too?

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cmd/cmd_analysis.c
Comment (b1llow) [inline librz/core/cmd/cmd_analysis.c]: @XVilka
We also keep some aliases, if we delete these aliases too, then we can delete `help_msg_ag`

File: librz/include/rz_core.h
Comment (ret2libc) [inline librz/include/rz_core.h:816]: I'd add also an API that takes a `RzCoreGraphType` as argument and prints the proper graph based on that. This way the cmd handlers can just use that API.

File: librz/core/cmd_descs/cmd_analysis.yaml
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_analysis.yaml]: you could use the CHOICE type and specify the choices as well, so that users can also more easily autocomplete.

File: librz/include/rz_core.h
Comment (XVilka) [reply librz/include/rz_core.h:816]: It should be an IPI though since there's no use of API function that prints something.

File: librz/core/cmd_descs/cmd_analysis.yaml
Comment (b1llow) [reply librz/core/cmd_descs/cmd_analysis.yaml]: Found a bug in CHOICE

image
image

Comment (b1llow) [reply librz/core/cmd_descs/cmd_analysis.yaml]: Always get an `Address boundary error`, when trying to trigger autocompletion

File: librz/include/rz_core.h
Comment (b1llow) [reply librz/include/rz_core.h:816]: Use `rz_core_graph_print`

File: librz/core/cmd_descs/cmd_analysis.yaml
Comment (XVilka) [reply librz/core/cmd_descs/cmd_analysis.yaml]: @ret2libc could you please take a look at this when you have time?

Comment (ret2libc) [reply librz/core/cmd_descs/cmd_analysis.yaml]: Will check

**Comment by XVilka** [comment]
@imbillow please rebase on top of the latest `dev`

**Review by XVilka — APPROVED** [review]
Please add a few tests with the full parameter names. Apart from that looks good for a first iteration, imho.

File: librz/core/cmd/cmd_analysis.c
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: not sure if this should be here or not.

File: librz/core/cmd/cmd_analysis.c
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]:
```suggestion
	char short_opt = strlen(x) == 1 ?  x[0] : 0;

	if (strncmp(x, "ascii", sizeof("ascii")) == 0 || RZ_STR_ISEMPTY(x) || short_opt == ' ') {
```

Comment (wargio) [reply librz/core/cmd/cmd_analysis.c]: can you implement something like this for the short options?

File: librz/core/cmd/cmd_analysis.c
Comment (b1llow) [reply librz/core/cmd/cmd_analysis.c]: I think it's needed, it looks like rzshell doesn't detect the value of choices

image

File: librz/core/cmd/cmd_analysis.c
Comment (wargio) [inline librz/core/cmd/cmd_analysis.c]: also, do not use `strncmp`; `strcmp` is enough.

**Review by wargio — CHANGES_REQUESTED** [review]
.

File: librz/core/cmd/cmd_analysis.c
Comment (b1llow) [reply librz/core/cmd/cmd_analysis.c]: Yes, but it can put `json-disasm` before `json`

**Comment by XVilka** [comment]
```

[XX] db/cmd/cmd_ag agd multiple formats
RZ_NOPLUGINS=1 /home/runner/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'aa 2> /dev/null
agd ascii main
agd cmd main
agd dot main
agd gml main
agd json main
agd json_disasm main
agd sdb main
agd tiny main
' bins/elf/hello_world
-- stdout
--- expected
+++ actual
@@ -100,3 +100,7 @@
  ()
 ()
  ()
+WARNING: rz_diff_levenstein_distance: assertion 'a && b' failed (line 69)
+WARNING: rz_diff_levenstein_distance: assertion 'a && b' failed (line 69)
+WARNING: rz_diff_levenstein_distance: assertion 'a && b' failed (line 69)
+WARNING: rz_diff_levenstein_distance: assertion 'a && b' failed (line 69)

-- stderr
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
```
https://github.com/rizinorg/rizin/runs/7908928663?check_suite_focus=true#step:18:251
```
C:\projects\rizin\librz\core\cmd\cmd_analysis.c(7244): warning C4244: 'function': conversion from 'double' to 'int', possible loss of data
```
AppVeyor:
```
[XX] C:\projects\rizin\test\db\cmd\cmd_ag agd multiple formats
ANSICON=1 RZ_NOPLUGINS=1 C:\projects\rizin\rizin-vs2022_64-v0.5.0\bin\rizin.exe -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'aa 2> nul
agd ascii main
agd cmd main
agd dot main
agd gml main
agd json main
agd json_disasm main
agd sdb main
agd tiny main
' bins/elf/hello_world
-- stdout
--- expected
+++ actual
@@ -33,9 +33,9 @@
  "0x000007aa" [fillcolor="lightgray",color="black", fontname="       ; DATA XREF from entry0 @ 0x6bd\lint main (int argc, char **argv, char **envp);\l     ; var int64_t var_20h @ rbp-0x20\l     ; var int64_t var_1ch @ rbp-0x1c\l     ; var int64_t var_18h @ rbp-0x18\l     ; var int64_t var_10h @ rbp-0x10\l     ; var int64_t var_8h @ rbp-0x8\l     push rbp\l     mov rbp, rsp\l     sub rsp, 0x20\l     ; 0x8d4\l     ; "Hello"\l     lea rax, str.Hello\l     mov qword [var_18h], rax\l     ; 0x8da\l     ; " r2-folks"\l     lea rax, str.r2_folks\l     mov qword [var_10h], rax\l     mov rax, qword [var_18h]\l     mov rdi, rax\l     ; size_t strlen(const char *s)\l     call sym.imp.strlen\l     mov dword [var_20h], eax\l     mov rax, qword [var_10h]\l     mov rdi, rax\l     ; size_t strlen(const char *s)\l     call sym.imp.strlen\l     mov dword [var_1ch], eax\l     mov edx, dword [var_20h]\l     mov eax, dword [var_1ch]\l     add eax, edx\l     add eax, 1\l     cdqe\l     mov rdi, rax\l     ; void *malloc(size_t size)\l     call sym.imp.malloc\l     mov qword [var_8h], rax\l     cmp qword [var_8h], 0\l     je 0x844\l", label="Courier", URL="main/0x000007aa"]
  "0x00000844" [fillcolor="lightgray",color="black", fontname="     nop\l     leave\l     ret\l", label="Courier", URL="main/0x00000844"]
  "0x00000806" [fillcolor="lightgray",color="black", fontname="     mov rdx, qword [var_18h]\l     mov rax, qword [var_8h]\l     mov rsi, rdx\l     mov rdi, rax\l     ; char *strcpy(char *dest, const char *src)\l     call sym.imp.strcpy\l     mov rdx, qword [var_10h]\l     mov rax, qword [var_8h]\l     mov rsi, rdx\l     mov rdi, rax\l     ; char *strcat(char *s1, const char *s2)\l     call sym.imp.strcat\l     mov rax, qword [var_8h]\l     mov rdi, rax\l     ; int puts(const char *s)\l     call sym.imp.puts\l     mov rax, qword [var_8h]\l     mov rdi, rax\l     ; void free(void *ptr)\l     call sym.imp.free\l", label="Courier", URL="main/0x00000806"]
-        "0x000007aa" -> "0x00000844" [color="#13a10e"];
-        "0x000007aa" -> "0x00000806" [color="#c50f1f"];
-        "0x00000806" -> "0x00000844" [color="#0037da"];
+        "0x000007aa" -> "0x00000844" [color="#008000"];
+        "0x000007aa" -> "0x00000806" [color="#800000"];
+        "0x00000806" -> "0x00000844" [color="#000080"];
 }
 graph
 [
[**]              C:\projects\rizin\test\db\cmd\cmd_agC     4300 OK       125 BR        0 XX       19 FX
[XX] C:\projects\rizin\test\db\cmd\cmd_ag agf multiple formats
ANSICON=1 RZ_NOPLUGINS=1 C:\projects\rizin\rizin-vs2022_64-v0.5.0\bin\rizin.exe -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'aa 2> nul
agf ascii
agf cmd
agf dot
agf gml
agf json
agf json_disasm
agf sdb
agf tiny
' bins/elf/hello_world
-- stdout
--- expected
+++ actual
@@ -31,7 +31,7 @@
 	graph [bgcolor=azure fontsize=8 fontname="Courier" splines="ortho"];
 	node [fillcolor=gray style=filled shape=box];
 	edge [arrowhead="normal"];
-	"0x000006a0" [URL="entry0/0x000006a0", fillcolor="white",color="#767676", fontname="Courier",label="  ;-- section..text:\l  ;-- .text:\l  ;-- _start:\lentry0 (int64_t arg3);\l; arg int64_t arg3 @ rdx\l0x000006a0      xor   ebp, ebp                                         ; [13] -r-x section size 546 named .text\l0x000006a2      mov   r9, rdx                                          ; arg3\l0x000006a5      pop   rsi\l0x000006a6      mov   rdx, rsp\l0x000006a9      and   rsp, 0xfffffffffffffff0\l0x000006ad      push  rax\l0x000006ae      push  rsp\l0x000006af      lea   r8, [sym.__libc_csu_fini]                        ; 0x8c0\l0x000006b6      lea   rcx, [sym.__libc_csu_init]                       ; 0x850 ; "AWAVA\\x89\\xffAUATL\\x8d%.\\x05 "\l0x000006bd      lea   rdi, [main]                                      ; 0x7aa\l0x000006c4      call  qword [reloc.__libc_start_main]                  ; [reloc.__libc_start_main:8]=0x201058 reloc.target.__libc_start_main ; "X\\x10 "\l"]
+	"0x000006a0" [URL="entry0/0x000006a0", fillcolor="white",color="#808080", fontname="Courier",label="  ;-- section..text:\l  ;-- .text:\l  ;-- _start:\lentry0 (int64_t arg3);\l; arg int64_t arg3 @ rdx\l0x000006a0      xor   ebp, ebp                                         ; [13] -r-x section size 546 named .text\l0x000006a2      mov   r9, rdx                                          ; arg3\l0x000006a5      pop   rsi\l0x000006a6      mov   rdx, rsp\l0x000006a9      and   rsp, 0xfffffffffffffff0\l0x000006ad      push  rax\l0x000006ae      push  rsp\l0x000006af      lea   r8, [sym.__libc_csu_fini]                        ; 0x8c0\l0x000006b6      lea   rcx, [sym.__libc_csu_init]                       ; 0x850 ; "AWAVA\\x89\\xffAUATL\\x8d%.\\x05 "\l0x000006bd      lea   rdi, [main]                                      ; 0x7aa\l0x000006c4      call  qword [reloc.__libc_start_main]                  ; [reloc.__libc_start_main:8]=0x201058 reloc.target.__libc_start_main ; "X\\x10 "\l"]
 }
 graph
 [
-- stderr
Interactive graph mode requires scr.interactive=true.
```
https://ci.appveyor.com/project/rizinorg/rizin/builds/44513693/job/b73fkw1yshhcaqv8#L4582

File: librz/core/cmd_descs/cmd_analysis.yaml
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_analysis.yaml:1001]: if you want you can use `details_cb: my_cb_function` (look at `rz_hash_bang_details_cb`) and for the choices below you can use another callback (with `choices_cb: my_other_cb_function`, see `rz_cmd_info_demangle_lang_choices`), so that you don't have to repeat the same values everywhere and you avoid repetition.

It is fine as is anyway... Just wanted to share what can be done with rzshell.

File: librz/core/cmd/cmd_analysis.c
Comment (wargio) [reply librz/core/cmd/cmd_analysis.c]: @ret2libc what do you think

File: librz/core/canalysis.c
Comment (ret2libc) [inline librz/core/canalysis.c]: I'd remove this... `df` works in visual mode, but not elsewhere. Also, command names could change in the future.

File: librz/core/cmd/cmd_analysis.c
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]:
```suggestion
	RzList *list = rz_str_split_duplist(dotPath, RZ_SYS_DIR, true);
```

File: librz/core/cmd/cmd_analysis.c
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]:
```suggestion
	static const bool callgraphs[] = {
```
does this work? I think it's better than defining a magic value

File: librz/core/cmd/cmd_analysis.c
Comment (ret2libc) [reply librz/core/cmd/cmd_analysis.c]: i think it should be handled differently, without `rz_warn_if_reached`, but yes, choices are not enforced at rzshell level, they are just suggested (for now at least..)

File: librz/core/cmd/cmd_analysis.c
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]: why `strncmp`? I think `strcmp` makes more sense, as we don't want to recognize 'asciiXXXXXX` as `ascii` option.

File: librz/core/cmd/cmd_analysis.c
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]: same here, use `strcmp`.

File: librz/core/cmd/cmd_analysis.c
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]: handle this case without the `rz_warn_*`.

File: librz/core/cmd_descs/cmd_analysis.yaml
Comment (b1llow) [reply librz/core/cmd_descs/cmd_analysis.yaml:1001]: But using details_cb here does not seem to have obvious advantages, but requires a lot of handwritten c code.

File: librz/core/cmd_descs/cmd_descs.h
Comment (XVilka) [inline librz/core/cmd_descs/cmd_descs.h:188]: @ret2libc seems generator produces duplicate items in case of `choices_cb`

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
@imbillow, why did you remove some tests?

**Comment by b1llow** [comment]
> @imbillow, why did you remove some tests?

Because the processing of `rz_core_analysis_graph` is very problematic, but I don't want to solve it in this PR, so temporarily delete these tests.

**Comment by XVilka** [comment]
@imbillow then please mark them "BROKEN=1" instead of removing and open an issue so they aren't forgotten.

**Comment by b1llow** [comment]
https://github.com/rizinorg/rizin/issues/2938

OK

**Review by XVilka — APPROVED** [review]

File: librz/diff/distance.c
Comment (ret2libc) [inline librz/diff/distance.c]: why? if it's NONNULL then rz_return shall be used and the caller should be fixed.

File: librz/core/cmd_descs/cmd_descs.h
Comment (ret2libc) [reply librz/core/cmd_descs/cmd_descs.h:188]: If it does not produce an error right now, I will create an issue and address it in a few days.

File: librz/diff/distance.c
Comment (b1llow) [reply librz/diff/distance.c]: `rz_return_val_if_fail` will print a warn message. It has no problem in most cases.
But https://github.com/rizinorg/rizin/runs/7917197046?check_suite_focus=true

It looks like stderr is mixed into stdout, in some tests

image

File: librz/core/cmd_descs/cmd_descs.h
Comment (XVilka) [reply librz/core/cmd_descs/cmd_descs.h:188]: OK: https://github.com/rizinorg/rizin/issues/2946

File: librz/diff/distance.c
Comment (XVilka) [reply librz/diff/distance.c]: I suggest also creating an issue for this to not block this PR. What do you think @ret2libc?

Comment (ret2libc) [reply librz/diff/distance.c]: Just check the parameters on the caller side. There are some new places (or some changes in this PR which trigger this) that calls this function with NULLs when they should not be.
I think it's better to "fix" it here, considering it was not happening before.

Comment (wargio) [reply librz/diff/distance.c]: this is wrong, that function must have non zero values.

**Comment by XVilka** [comment]
@ret2libc please take a look again and let's merge this if you don't have any big objections. Further graph improvements can be done in separate PRs.

**Review by ret2libc — APPROVED** [review]

**Pull request merged**
